[SPARK-14459] [SQL] Detect relation partitioning and adjust the logical plan

rdblue opened the pull request:
## What changes were proposed in this pull request?

This detects a relation's partitioning and adds checks to the analyzer.
If an InsertIntoTable node has no partitioning, it is replaced by the
relation's partition scheme and input columns are correctly adjusted,
placing the partition columns at the end in partition order. If an
InsertIntoTable node has partitioning, it is checked against the table's
reported partitions.

These changes required adding a PartitionedRelation trait to the catalog
interface because Hive's MetastoreRelation doesn't extend
CatalogRelation.

This commit also includes a fix to InsertIntoTable's resolved logic,
which now detects that all expected columns are present, including
dynamic partition columns. Previously, the number of expected columns
was not checked and resolved was true if there were missing columns.
## How was this patch tested?

This adds new tests to the InsertIntoTableSuite that are fixed by this PR.

rdblue committed:
SPARK-14459: Detect relation partitioning and adjust the logical plan to match.

This detects a relation's partitioning and adds checks to the analyzer.
If an InsertIntoTable node has no partitioning, it is replaced by the
relation's partition scheme and input columns are correctly adjusted,
placing the partition columns at the end in partition order. If an
InsertIntoTable node has partitioning, it is checked against the table's
reported partitions.

These changes required adding a PartitionedRelation trait to the catalog
interface because Hive's MetastoreRelation doesn't extend
CatalogRelation.

This commit also includes a fix to InsertIntoTable's resolved logic,
which now detects that all expected columns are present, including
dynamic partition columns. Previously, the number of expected columns
was not checked and resolved was true if there were missing columns.

rdblue commented:
@liancheng, can you look at this? It looks like you're familiar with the SQL/Hive code.

AmplabJenkins commented:
Can one of the admins verify this patch?

marmbrus commented:
ok to test

SparkQA commented:
**[Test build #55231 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55231/consoleFull)** for PR 12239 at commit [`5e10c7c`](https://github.com/apache/spark/commit/5e10c7c83ae9db0c2f79fbaa16457d954450291e).

SparkQA commented:
**[Test build #55231 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55231/consoleFull)** for PR 12239 at commit [`5e10c7c`](https://github.com/apache/spark/commit/5e10c7c83ae9db0c2f79fbaa16457d954450291e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55231/
Test FAILed.

rdblue committed:
SPARK-14459: Update partition spec validation test.

This test expected to fail the strict partition check, but with support
for table partitioning in the analyzer the problem is caught sooner and
has a better error message. The message now complains that the
partitioning doesn't match rather than strict mode, which wouldn't help.

SparkQA commented:
**[Test build #55399 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55399/consoleFull)** for PR 12239 at commit [`7c31b73`](https://github.com/apache/spark/commit/7c31b73b2f5471bd4e4764f2132f056584db5890).

rdblue commented:
@marmbrus, I fixed the failing test. The problem was a query that didn't supply a value for one of the partitions. This commit actually gives a better error message, but it is an AnalysisException rather than a SparkException. Thanks for taking a look at this.

SparkQA commented:
**[Test build #55399 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55399/consoleFull)** for PR 12239 at commit [`7c31b73`](https://github.com/apache/spark/commit/7c31b73b2f5471bd4e4764f2132f056584db5890).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55399/
Test FAILed.

rdblue commented:
Looks like I accidentally pushed a commit I didn't intend to. I've fixed that, but the test failed. This is ok to test (again).

SparkQA commented:
**[Test build #55403 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55403/consoleFull)** for PR 12239 at commit [`2a807a9`](https://github.com/apache/spark/commit/2a807a97cf793a480dbaf172789a315624db44b9).

SparkQA commented:
**[Test build #55403 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55403/consoleFull)** for PR 12239 at commit [`2a807a9`](https://github.com/apache/spark/commit/2a807a97cf793a480dbaf172789a315624db44b9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55403/
Test FAILed.

rdblue commented:
I just pushed a fix for the last test failure. Should be ready to test again.

SparkQA commented:
**[Test build #55533 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55533/consoleFull)** for PR 12239 at commit [`6491788`](https://github.com/apache/spark/commit/6491788a3419b091dfd6ca2a14eebb9b0ec33635).

SparkQA commented:
**[Test build #55533 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55533/consoleFull)** for PR 12239 at commit [`6491788`](https://github.com/apache/spark/commit/6491788a3419b091dfd6ca2a14eebb9b0ec33635).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class OneRowRelation(outputAttrs: Seq[Attribute]) extends LeafNode `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55533/
Test FAILed.

rdblue committed:
SPARK-14459: Fix OneRowTable error message.

OneRowTable doesn't expose its output columns because it is a singleton.
The more strict checks in InsertIntoTable's resolution is causing this
to fail. This commit fixes the problem by catching the case where a
table doesn't define its outputs and considers the table resolved.

SparkQA commented:
**[Test build #55563 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55563/consoleFull)** for PR 12239 at commit [`aad3eb5`](https://github.com/apache/spark/commit/aad3eb5c33ac0324970671ede70736360d60f0df).

SparkQA commented:
**[Test build #55563 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55563/consoleFull)** for PR 12239 at commit [`aad3eb5`](https://github.com/apache/spark/commit/aad3eb5c33ac0324970671ede70736360d60f0df).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55563/
Test PASSed.

rxin commented:
cc @liancheng and @cloud-fan

SparkQA commented:
**[Test build #56387 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56387/consoleFull)** for PR 12239 at commit [`2247648`](https://github.com/apache/spark/commit/224764828d58b35168b405fe1cc46b628ac1afc9).

SparkQA commented:
**[Test build #56387 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56387/consoleFull)** for PR 12239 at commit [`2247648`](https://github.com/apache/spark/commit/224764828d58b35168b405fe1cc46b628ac1afc9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56387/
Test PASSed.

cloud-fan commented:
hi @rxin , are we going to make `MetastoreRelation` also a `CatalogTable`?

rxin commented:
I think so - the current MetastoreRelation doesn't make a lot of sense.

rdblue commented:
@rxin, @cloud-fan, this PR works for both cases when the table is resolved. I think making MetastoreRelation a CatalogTable would certainly improve things, but it looks like that is a long-term plan rather than something that should be solved before this goes in.

By the way, my other PR to align writes more with user expectations, #12313, also works toward making Hive behave like other relations. It consolidates the pre-write checks and casts and makes it much more user-friendly to write into any relation.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -414,8 +414,42 @@ class Analyzer(
+            if (parts.keys.nonEmpty) {
+              // the query's partitioning must match the table's partitioning
+              // this is set for queries like: insert into ... partition (one = "a", two = <expr>)
+              if (tablePartitionNames.size != parts.keySet.size) {
```
why do we only check size here?

cloud-fan replied:
why do we only check size here?

rdblue replied:
In this case, the user has called `partitionBy(...)` to set the partition columns by hand. Those don't have to match by name and we can insert casts during the pre-insert checks because we trust that the user gets it right. The only thing valuable to check is that the user got the number of partitions right as a sanity check.

cloud-fan replied:
I checked the `PreWriteCheck` rule, it does check if the user specified partition columns match the real ones. So I think we don't need to check here but just add a comment to say that the `PreWriteCheck` will do this check.

rdblue replied:
Works for me.

rdblue replied:
Done.

rdblue replied:
Removing this check broke 2 tests because the pre-insert code isn't correctly checking it. My follow-up PR fixes and merges the pre-inserts, so I've added it back in this PR with a TODO to remove it. I'll remove it when I rebase the follow-up on this.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -414,8 +414,42 @@ class Analyzer(
+                tablePartitionNames.contains(attr.name)
+              }
+              // All partition columns are dynamic because this InsertIntoTable had no partitioning
+              val partColumns = tablePartitionNames.map { name =>
```
when will the `partColumns` different from `inputPartCols`? Seems neve?

rdblue replied:
This does two things:
# Reorders the partition columns to match the table. There's a test case for this and it catches a small bug where the wrong data column is used when the input partition columns don't match the order of the table's partitioning.
# If the column is missing from the data to be written, this will throw an exception.

cloud-fan replied:
makes sense.

rdblue commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertIntoHiveTableSuite.scala:
```diff
@@ -259,4 +261,78 @@ class InsertIntoHiveTableSuite extends QueryTest with TestHiveSingleton with Bef
+    sqlContext.conf.unsetConf("hive.exec.dynamic.partition.mode")
+  }
+
+  test("Detect table partitioning with correct partition order") {
```
@cloud-fan, this is the test that checks partition columns are in the correct order for the table (since Hive assumes they are when writing).

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertIntoHiveTableSuite.scala:
```diff
@@ -259,4 +261,78 @@ class InsertIntoHiveTableSuite extends QueryTest with TestHiveSingleton with Bef
   }
+
+  test("Reject partitioning that does not match table") {
+    sqlContext.setConf("hive.exec.dynamic.partition.mode", "nonstrict")
```
please use `withConf`, so that these conf will be cleaned after test

rdblue replied:
Fixed.

cloud-fan commented:
cc @liancheng to take another look

rdblue committed:
SPARK-14459: Changes for review feedback.

rdblue commented:
@cloud-fan, I've rebased on master and fixed the two things you pointed out. Let me know if there's anything else. Thanks for reviewing!

SparkQA commented:
**[Test build #56710 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56710/consoleFull)** for PR 12239 at commit [`d8aab8a`](https://github.com/apache/spark/commit/d8aab8a7503a72e52623b2b079ceee46e963904a).

SparkQA commented:
**[Test build #56710 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56710/consoleFull)** for PR 12239 at commit [`d8aab8a`](https://github.com/apache/spark/commit/d8aab8a7503a72e52623b2b079ceee46e963904a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56710/
Test FAILed.

SparkQA commented:
**[Test build #56737 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56737/consoleFull)** for PR 12239 at commit [`d86161c`](https://github.com/apache/spark/commit/d86161c073d876501e82a1ff9180c9531baae48d).

SparkQA commented:
**[Test build #56744 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56744/consoleFull)** for PR 12239 at commit [`023d656`](https://github.com/apache/spark/commit/023d6566a0b4adbe64f113f9d542048b020ae2d6).

rdblue commented:
@cloud-fan, withSQLConf and withHadoopConf weren't working so I added withSessionConf to set the partitioning property correctly.

SparkQA commented:
**[Test build #56746 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56746/consoleFull)** for PR 12239 at commit [`2c4d7af`](https://github.com/apache/spark/commit/2c4d7afac8aae2686c5d613c13e87d2230ae5c09).

SparkQA commented:
**[Test build #56737 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56737/consoleFull)** for PR 12239 at commit [`d86161c`](https://github.com/apache/spark/commit/d86161c073d876501e82a1ff9180c9531baae48d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56737/
Test FAILed.

SparkQA commented:
**[Test build #56744 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56744/consoleFull)** for PR 12239 at commit [`023d656`](https://github.com/apache/spark/commit/023d6566a0b4adbe64f113f9d542048b020ae2d6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56744/
Test FAILed.

SparkQA commented:
**[Test build #56746 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56746/consoleFull)** for PR 12239 at commit [`2c4d7af`](https://github.com/apache/spark/commit/2c4d7afac8aae2686c5d613c13e87d2230ae5c09).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56746/
Test FAILed.

SparkQA commented:
**[Test build #56755 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56755/consoleFull)** for PR 12239 at commit [`82add06`](https://github.com/apache/spark/commit/82add06177c6b730459aea5eb7e277a0615147fd).

SparkQA commented:
**[Test build #56755 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56755/consoleFull)** for PR 12239 at commit [`82add06`](https://github.com/apache/spark/commit/82add06177c6b730459aea5eb7e277a0615147fd).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56755/
Test FAILed.

rdblue commented:
I'm not sure what's going on with the MiMa tests. Looks unrelated to my PR. I think all of the tests are passing again so this should be fine. Jenkins should retest if possible.

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #56776 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56776/consoleFull)** for PR 12239 at commit [`82add06`](https://github.com/apache/spark/commit/82add06177c6b730459aea5eb7e277a0615147fd).

SparkQA commented:
**[Test build #56776 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56776/consoleFull)** for PR 12239 at commit [`82add06`](https://github.com/apache/spark/commit/82add06177c6b730459aea5eb7e277a0615147fd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56776/
Test PASSed.

SparkQA commented:
**[Test build #56930 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56930/consoleFull)** for PR 12239 at commit [`9016f7d`](https://github.com/apache/spark/commit/9016f7d5b357015d30db31cfdcef88b7e783a852).

SparkQA commented:
**[Test build #56930 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56930/consoleFull)** for PR 12239 at commit [`9016f7d`](https://github.com/apache/spark/commit/9016f7d5b357015d30db31cfdcef88b7e783a852).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56930/
Test PASSed.

rdblue commented:
I noticed that there was a conflict so I rebased on master. Tests are still passing.

SparkQA commented:
**[Test build #57137 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57137/consoleFull)** for PR 12239 at commit [`39fe90c`](https://github.com/apache/spark/commit/39fe90cdbd4690c0c48ed73ac00fd694a9976908).

SparkQA commented:
**[Test build #57137 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57137/consoleFull)** for PR 12239 at commit [`39fe90c`](https://github.com/apache/spark/commit/39fe90cdbd4690c0c48ed73ac00fd694a9976908).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57137/
Test FAILed.

SparkQA commented:
**[Test build #57138 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57138/consoleFull)** for PR 12239 at commit [`43a6e8b`](https://github.com/apache/spark/commit/43a6e8b4f2d90bc3f601de65a8a3bcd352ca52f5).

SparkQA commented:
**[Test build #57138 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57138/consoleFull)** for PR 12239 at commit [`43a6e8b`](https://github.com/apache/spark/commit/43a6e8b4f2d90bc3f601de65a8a3bcd352ca52f5).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57138/
Test FAILed.

SparkQA commented:
**[Test build #57144 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57144/consoleFull)** for PR 12239 at commit [`f3c3cdb`](https://github.com/apache/spark/commit/f3c3cdbb0571964fe5a4003fb080bf538a11af4e).

SparkQA commented:
**[Test build #57144 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57144/consoleFull)** for PR 12239 at commit [`f3c3cdb`](https://github.com/apache/spark/commit/f3c3cdbb0571964fe5a4003fb080bf538a11af4e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57144/
Test FAILed.

SparkQA commented:
**[Test build #57350 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57350/consoleFull)** for PR 12239 at commit [`2130db8`](https://github.com/apache/spark/commit/2130db80e6717060aad7f37f445cad6ec861252d).

SparkQA commented:
**[Test build #57350 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57350/consoleFull)** for PR 12239 at commit [`2130db8`](https://github.com/apache/spark/commit/2130db80e6717060aad7f37f445cad6ec861252d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57350/
Test PASSed.

rdblue commented:
@cloud-fan, I rebased on master to avoid the conflicts and tests are all passing. If you have a chance to take another look I'd appreciate it! I think this is about ready.

rxin commented:
Ryan - I've added this to the review list for the team.

liancheng commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -347,10 +347,23 @@ case class InsertIntoTable(
+      val (partitionColumns, dataColumns) = table.output
+          .partition(a => partition.keySet.contains(a.name))
+      Some(dataColumns ++ partitionColumns.takeRight(numDynamicPartitions))
+    }
```
Seems that we can omit the `if` condition and just use a `Seq[Attribute]` instead of `Option[Seq[Attribute]]` here?

rdblue replied:
@liancheng, I added a note about this below. The problem is that some relations return `Nil` as their output. I think this approach is the least intrusive solution because changing those relations is a large change. Please let me know what you think, and thanks for reviewing!

liancheng commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertIntoHiveTableSuite.scala:
```diff
@@ -212,4 +214,77 @@ class InsertIntoHiveTableSuite extends QueryTest with TestHiveSingleton with Bef
   }
+
+  test("Reject partitioning that does not match table") {
+    withSQLConf(("hive.exec.dynamic.partition.mode", "nonstrict")) {
```
I doubt whether this configuration is effective/respected in Spark. Not quite familiar with the newly refactored Spark configuration API. I guess it's probably not respected, and you don't need it to create tables partitioned by dynamic partitions only in Spark.

rdblue replied:
This is enforced by the dynamic partition writer container for Hive to match Hive's behavior. Since these tests don't have a static partition, it's needed or else the tests fail. I can change this, but I'm not sure what you would suggest.

liancheng replied:
Ah sorry, made a mistake here. I only checked that `CREATE TABLE` with dynamic partitions only doesn't require setting this flag, but this flag does affect the insertion below.

liancheng commented:
Hey @rdblue, glad to see you here :) Sorry for the late reply, I somehow missed the initial message.

This mostly LGTM except for two minor comments. Thanks!

SparkQA commented:
**[Test build #57909 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57909/consoleFull)** for PR 12239 at commit [`6aa17f4`](https://github.com/apache/spark/commit/6aa17f4b65b7d19ec0ba92a3a6ce9246012055cc).

SparkQA commented:
**[Test build #57909 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57909/consoleFull)** for PR 12239 at commit [`6aa17f4`](https://github.com/apache/spark/commit/6aa17f4b65b7d19ec0ba92a3a6ce9246012055cc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57909/
Test FAILed.

SparkQA commented:
**[Test build #57919 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57919/consoleFull)** for PR 12239 at commit [`2130db8`](https://github.com/apache/spark/commit/2130db80e6717060aad7f37f445cad6ec861252d).

rdblue commented:
@liancheng, I originally had the logic you suggest for the expected columns calculation. But, there were [test failures](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57909/consoleFull) because `OneRowRelation` reports its output columns as `Nil` and will never match the input. Changing that introduced a huge number of test failures where other code depended on that Nil, so I think the best way forward is to catch the case where the output table has no columns and consider the node resolved.

The down side to this approach is that tables that actually have no columns would be considered resolved, but we will catch this case in later processing. For example, the `OneRowRelation` test is actually looking for a helpful error message that you can't write to an RDD-based relation.

rdblue commented:
@liancheng and @rxin, thank you for looking at this!

SparkQA commented:
**[Test build #57919 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/57919/consoleFull)** for PR 12239 at commit [`2130db8`](https://github.com/apache/spark/commit/2130db80e6717060aad7f37f445cad6ec861252d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/57919/
Test PASSed.

liancheng commented:
I'm merging this to master and branch-2.0. Thanks for fixing this!

rdblue commented:
Thanks for reviewing this, @cloud-fan and @liancheng!
